src/PartitionLoops.cpp

```diff
@@ -0,0 +1,924 @@
+            body = mutate(op->body);
+            scope.pop(op->name);
+        } else {
+            body = mutate(op->body);
+        }
+
+        if (value.same_as(op->value) && body.same_as(op->body)) {
+            result = op;
+        } else {
+            result = LetOrLetStmt::make(op->name, value, body);
+        }
+    }
+
+    void visit(const Let *op) {
+        visit_let(op, expr);
+    }
+
+    void visit(const LetStmt *op) {
+        visit_let(op, stmt);
+    }
+
+public:
+    BoundsBasedSimplify(const string &loop_var, Expr min, Expr max) {
+        scope.push(loop_var, Interval(min, max));
+    }
+};
+
+class PartitionLoops : public IRMutator {
+    using IRMutator::visit;
+
+    void visit(const For *op) {
+
+        Stmt body = mutate(op->body);
+
+        // Conservatively only apply this optimization at one loop
+        // level. Comment this out to allow full expansion.
+        if (!body.same_as(op->body)) {
+            stmt = For::make(op->name, op->min, op->extent,
+                             op->for_type, op->device_api, body);
+            return;
+        }
+
+        FindSteadyState f(op->name);
+        Stmt simpler_body = f.mutate(body);
+        // Ask for the start and end of the steady state.
+        Expr min_steady = f.min_steady_val();
+        Expr max_steady = f.max_steady_val();
+
+        if (min_steady.defined() || max_steady.defined()) {
+            debug(3) << "\nOld body: " << body << "\n";
+
+            // They're undefined if there's no prologue or epilogue.
+            bool make_prologue = min_steady.defined();
+            bool make_epilogue = max_steady.defined();
+
+            // Accrue a stack of let statements defining the steady state start and end.
+            vector<pair<string, Expr> > lets;
+
+            // The steady state was simplified by
+            // FindSteadyState. It's not safe to simplify the prologue
+            // and epilogue at the same time - FindSteadyState uses
+            // already found simplifications to make new ones,
+            // including potentially using a simplification of the
+            // prologue end to assist the epilogue start. Instead we
+            // do a bounds-based simplification of the prologue and
+            // epilogue as a later pass.
+            //
+            // Note that the bounds we use are actually slightly
+            // conservative. E.g. we claim the prologue can run up
+            // till min_steady, where it actually runs to
+            // clamp(min_steady, loop_min, loop_max). Fortunately, the
+            // case where our bound is incorrect is the case where the
+            // prologue or epilogue have extent zero, so the incorrect
+            // simplification doesn't matter.
+            BoundsBasedSimplify simplify_prologue(op->name, op->min, min_steady.defined() ? min_steady - 1 : Expr());
+            BoundsBasedSimplify simplify_epilogue(op->name, max_steady, op->min + op->extent - 1);
```

Comment:
Oh wow, I forgot to delete a bunch of dead code. Good catch.

On Wed, Mar 25, 2015 at 2:47 PM, Kree Cole-McLaughlin <
notifications@github.com> wrote:

> In src/PartitionLoops.cpp
> <https://github.com/halide/Halide/pull/713#discussion_r27169810>:
>
> > +            // and epilogue at the same time - FindSteadyState uses
> > +            // already found simplifications to make new ones,
> > +            // including potentially using a simplification of the
> > +            // prologue end to assist the epilogue start. Instead we
> > +            // do a bounds-based simplification of the prologue and
> > +            // epilogue as a later pass.
> > +            //
> > +            // Note that the bounds we use are actually slightly
> > +            // conservative. E.g. we claim the prologue can run up
> > +            // till min_steady, where it actually runs to
> > +            // clamp(min_steady, loop_min, loop_max). Fortunately, the
> > +            // case where our bound is incorrect is the case where the
> > +            // prologue or epilogue have extent zero, so the incorrect
> > +            // simplification doesn't matter.
> > +            BoundsBasedSimplify simplify_prologue(op->name, op->min, min_steady.defined() ? min_steady - 1 : Expr());
> > +            BoundsBasedSimplify simplify_epilogue(op->name, max_steady, op->min + op->extent - 1);
>
> It looks like these objects are no longer used.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/713/files#r27169810>.
>